test/test_nested_tensor_masking.py

```diff
@@ -39,312 +45,995 @@ def test_gen_nested_tensor(self):
+        self.assertEqual(tensor, torch.tensor([[1]], dtype=torch.bfloat16))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+
+        tensor, mask = a.to_tensor_mask(mask_dim=1)
+        self.assertEqual(tensor, torch.tensor([[1]], dtype=torch.bfloat16))
+        self.assertEqual(mask, torch.tensor([True]))
+        check_tensor_options(self, tensor, a)
+
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
+        self.assertEqual(tensor, torch.tensor([[1]], dtype=torch.bfloat16))
+        self.assertEqual(mask, torch.tensor([[True]]))
+        check_tensor_options(self, tensor, a)
+
+        self.assertRaisesRegex(RuntimeError, "Mask dimention is bigger than nested dimention of a nested tensor.", lambda: a.to_tensor_mask(mask_dim=3))
+
+    #TODO once .to_list() bug fixed
+    def test_multi_scalar(self):
+        # TODO: add test cases
+        #a = nt.nested_tensor([
+        #        torch.tensor(1),
+        #        torch.tensor(2),
+        #        torch.tensor(3)
+        #    ])
+        #tensor, mask = a.to_tensor_mask()
+
+        a = nt.nested_tensor([
+                nt.nested_tensor([
+                    torch.tensor(1),
+                    torch.tensor(2),
+                    torch.tensor(3)
+                ])
+            ])
+
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[1, 2, 3]]))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
+        self.assertEqual(tensor, torch.tensor([[1, 2, 3]]))
+        self.assertEqual(mask, torch.tensor([[True, True, True]]))
+        check_tensor_options(self, tensor, a)
+
+        self.assertRaisesRegex(RuntimeError, "Mask dimention is bigger than nested dimention of a nested tensor.", lambda: a.to_tensor_mask(mask_dim=3))
+
+        a = nt.nested_tensor([
                 nt.nested_tensor([
-                    torch.tensor([[1, 2], 
-                                [3, 4]])
+                    torch.tensor(1)
                 ]),
-            ]),
-            nt.nested_tensor([
                 nt.nested_tensor([
-                    torch.tensor([[5, 6],
-                                [7, 8]])
+                    torch.tensor(2)
                 ]),
+                nt.nested_tensor([
+                    torch.tensor(3)
+                ])
+            ])
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[1], [2], [3]]))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
+        self.assertEqual(tensor, torch.tensor([[1], [2], [3]]))
+        self.assertEqual(mask, torch.tensor([[True], [True], [True]]))
+        check_tensor_options(self, tensor, a)
+
+    def test_scalar_and_empty_nt(self):
+        a = nt.nested_tensor([
```

Comment:
We expect this construction to fail in the future, right? Since the first entry is of type float (by default) and the second of type long.

Reply:
Yep